pkg/problemmetrics/problem_metrics.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2019 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package problemmetrics
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"k8s.io/node-problem-detector/pkg/util/metrics"
+)
+
+// GlobalProblemMetricsManager is a singleton of ProblemMetricsManager,
+// which should be used to manage all problem-converted metrics across all
+// problem daemons.
+var GlobalProblemMetricsManager *ProblemMetricsManager
+
+func init() {
+	GlobalProblemMetricsManager = NewProblemMetricsManagerOrDie()
+}
+
+// ProblemMetricsManager manages problem-converted metrics.
+// ProblemMetricsManager is thread-safe.
+type ProblemMetricsManager struct {
+	problemCounter         metrics.Int64MetricInterface
+	problemGauge           metrics.Int64MetricInterface
+	conditionToReason      map[string]string
+	conditionToReasonMutex sync.Mutex
+}
+
+func NewProblemMetricsManagerOrDie() *ProblemMetricsManager {
+	pmm := ProblemMetricsManager{}
+
+	var err error
+	pmm.problemCounter, err = metrics.NewInt64Metric(
+		"problem_counter",
+		"Number of times a specific type of problem have occurred.",
+		"1",
+		metrics.Sum,
+		[]string{"reason"})
+	if err != nil {
+		glog.Fatalf("Failed to create problem_counter metric: %v", err)
+	}
+
+	pmm.problemGauge, err = metrics.NewInt64Metric(
+		"problem_gauge",
+		"Whether a specific type of problem is affecting the node or not.",
+		"1",
+		metrics.LastValue,
+		[]string{"condition", "reason"})
+	if err != nil {
+		glog.Fatalf("Failed to create problem_gauge metric: %v", err)
+	}
+
+	pmm.conditionToReason = make(map[string]string)
+
+	return &pmm
+}
+
+// NewProblem reports a temporary problem that results in problem counter increment.
+func (pmm *ProblemMetricsManager) NewProblem(reason string) error {
+	if pmm.problemCounter == nil {
+		return errors.New("problem counter is being incremented before initialized.")
+	}
+
+	return pmm.problemCounter.Record(map[string]string{"reason": reason}, 1)
+}
+
+// SetCondition reports a permanent problem that result in a problem gauge value set.
+func (pmm *ProblemMetricsManager) SetCondition(condition string, reason string, value bool) error {
+	if pmm.problemGauge == nil {
+		return errors.New("problem gauge is being set before initialized.")
+	}
+
+	pmm.conditionToReasonMutex.Lock()
+	defer pmm.conditionToReasonMutex.Unlock()
+
+	lastReason := pmm.conditionToReason[condition]
+	if lastReason != "" {
```

Comment:
If you are checking if `condition` is in the map, it is better to use `ok`:
```
if lastReason, ok := pmm.conditionToReason[condition], ok {
    ...
}
```
instead of assuming `lastReason` is not empty string.

Reply:
Ah sorry. Don't know what was I thinking.